### ISSUE TITLE
Fluentd in_tail "unreadable" file causes "following tail of <file>" to stop and no logs pushed

### ISSUE THREAD

**Issue by salavessa** [fluent/fluentd#3614]
### Describe the bug

After a warning of an "unreadable" (likely due to rotation), no more logs were pushed (in_tail + pos_file).
Reloading config or restarting fluentd sorts the issue.
All other existing files being tracked continued to work as expected.

### To Reproduce

Not able to reproduce at will.

### Expected behavior

Logs to be pushed as usual after file rotation as fluentd recovers from the temporary "unreadable" file.

### Your Environment

```markdown
- Fluentd version: 1.14.4
- TD Agent version: 4.2.0
- Operating system: Ubuntu 20.04.3 LTS
- Kernel version: 5.11.0-1022-aws
```

### Your Configuration

```apache
<source>
  @type tail
  path /var/log/containers/*.log
  pos_file /var/log/es-containers.log.pos
  tag kubernetes.*
  <parse>
    @type regexp
    expression /^(?<time>[^ ]+) (?<stream>[^ ]+) (?<logtag>[^ ]+) (?<log>.+)$/
    time_key time
    time_format %Y-%m-%dT%H:%M:%S.%N%z
  </parse>
  read_from_head true
</source>
```

### Your Error Log

```shell
Relevant log entries with some context. When "detected rotation of ..." isn't followed by a "following tail of ..." then log file contents aren't being processed/pushed:

2022-02-01 01:26:33 +0000 [info]: #0 detected rotation of /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log; waiting 5 seconds
2022-02-01 01:26:33 +0000 [info]: #0 following tail of /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log
2022-02-01 01:32:53 +0000 [warn]: #0 /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log unreadable. It is excluded and would be examined next time.
2022-02-01 01:32:54 +0000 [info]: #0 detected rotation of /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log; waiting 5 seconds
2022-02-01 01:32:54 +0000 [info]: #0 following tail of /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log
2022-02-01 01:38:04 +0000 [info]: #0 detected rotation of /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log; waiting 5 seconds
2022-02-01 01:44:44 +0000 [info]: #0 detected rotation of /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log; waiting 5 seconds
2022-02-01 01:53:15 +0000 [info]: #0 detected rotation of /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log; waiting 5 seconds
[...]
---- after issuing a config reload (kill -SIGUSR2 <pid>) it starts to work fine again, i.e. "following tail of ..." 
2022-02-01 11:36:19 +0000 [info]: Reloading new config
2022-02-01 11:36:19 +0000 [info]: using configuration file: <ROOT>
[...]
2022-02-01 11:36:19 +0000 [info]: shutting down input plugin type=:tail plugin_id="object:c6c0"
[...]
2022-02-01 11:36:19 +0000 [info]: adding source type="tail"
2022-02-01 11:36:19 +0000 [info]: #0 shutting down fluentd worker worker=0
2022-02-01 11:36:19 +0000 [info]: #0 shutting down input plugin type=:tail plugin_id="object:c6c0"
[...]
2022-02-01 11:36:20 +0000 [info]: #0 restart fluentd worker worker=0
---- the entry below seems to be related with the actual underlying issue... the ruby object which stopped pushing logs has now been terminated as a new one was created
2022-02-01 11:36:20 +0000 [warn]: #0 /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log already exists. use latest one: deleted #<struct Fluent::Plugin::TailInput::Entry path="/var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log", pos=10740032, ino=1797715, seek=1530>
2022-02-01 11:36:20 +0000 [info]: #0 following tail of /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log
2022-02-01 11:37:30 +0000 [info]: #0 detected rotation of /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log; waiting 5 seconds
2022-02-01 11:37:30 +0000 [info]: #0 following tail of /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log
2022-02-01 11:43:20 +0000 [info]: #0 detected rotation of /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log; waiting 5 seconds
2022-02-01 11:43:20 +0000 [info]: #0 following tail of /var/log/containers/gateway-mtcw6_atom_pcms-gateway-e54ea9ee115f1127682b0d92e37404e4b1693c14edeea93e485f70b3110eecfc.log
```

### Additional context

This issue seems be related with #3586 but unfortunately I didn't check the `pos` file while the issue was happening so can't tell if it presented unexpected values for the failing file.

**Comment by mizunashi-mana** [comment]
We have encountered the same problem twice. Last month we upgraded fluentd from 1.13.3 (td-agent 4.2.0) to 1.14.3 (td-agent 4.3.0), and then we have got the problem.

In our case, we are tracking more than 3 log files with the tail plugin, and only the file with the highest log flow has this problem. From the metrics produced by prometheus plugin, the file position always fluctuates by at least 400000/sec and the rotation happens at least once every 10 minutes.

The input source is set as follows:

```
<source>
  @type tail
  format ...
  path /var/log/server.log
  pos_file /var/log/td-agent/server.log.pos
  tag input.server.log
</source>
```

I will also share the record of the investigation when the problem occurs. First, fluentd suddenly, without warning, stops tracking new rotated files and only logs the following in the stdout log.

```
detected rotation of /var/log/server.log; waiting 5 seconds
```

Normally, it would have been followed by an announcement that a new file was being tracked, as follows

``` .
The following tail of /var/log/server.log
```

However, we have not seen this announcement at all after the problem occurs. Also, when I view the file descriptors that fluentd is tracking, I see the following:

```
$ sudo ls "/proc/$(pgrep -f 'td-agent.*--under-supervisor')/fd" -l | grep server
lr-x------ 1 td-agent td-agent 64 Feb 5 01:41 49 -> /var/log/server.3.log
```

This is the rotated file, and we see that fluentd is still tracking the old file, not the new one. Also, looking at the position file, we see the following:

```
$ sudo cat /var/log/td-agent/server.log.pos 
/var/log/server.log 000000002001f2d0 000000003a000042
/var/log/server.log 0000000000000000 0000000000000000
```

As far as the situation is concerned, it seems that something triggers fluentd to stop tracking new files altogether during rotate. We have about 300 servers running 24/7 and have only had two problems in the past month, so it seems to be pretty rare for problems to occur. However, since we haven't had any problems with fluentd 1.13.3 for 6 months, it seems likely that there was some kind of regression in fluentd 1.13.3 -> 1.14.3.

We are planning to downgrade fluentd.

**Comment by ashie** [comment]
Memo: From v1.13.3 to v1.14.3, following changes are made to in_tail:

* #3467
* #3478
* #3500
* #3504
* #3541
* #3542

**Comment by mizunashi-mana** [comment]
We downgraded td-agent 4.3.0 (fluentd v1.14.3) to td-agent 4.2.0 (fluentd v1.13.3), and still have problems.
We broadened monitoring targets to 500 servers (approx 1500 files) from 200 servers (approx 400 files) as a result of this issue. Then, we found the 2 servers (2 files) having this issue after a week from td-agent downgraded.
Upgrading fluentd 1.13.3 -> 1.14.3 does not seem to be the cause, sorry.

It seemed that some of our log files were affected and some were not. The differences are as follows.

* rotation policy; the unaffected files were using a time based rotation policy (e.g. *.log-%Y%m%d-%s), while the affected files were using a fixed window policy (e.g. *.%i.log). The fixed window policy takes longer to rotate than the time based rotation policy (because it requires renaming all the previous files).
* rotation period; unaffected files are rotated in 30 minutes, while affected files are rotated in about 7 minutes. However, it may be just a matter of probability since the more rotations we do, the higher the probability that we will get this problem.

We are planning to change rotation policy and retry upgrading td-agent. And, we will comment some updates if we get.

**Comment by mizunashi-mana** [comment]
We enabled debug logging with fluentd 1.13.3 and encountered this issue. The state was same:

* fluentd had been reporting `detected rotation of /var/log/server.log; waiting 5 seconds` without `The following tail of /var/log/server.log`.
* fluentd tracked the old file descriptor.
* position file had two entries; a correct position and `/var/log/server.log 0000000000000000 0000000000000000`.
* `fluentd_tail_file_inode` metrics had been keeping the same inode from this issue occurred.

And, fluentd reported `Skip update_watcher because watcher has been already updated by other inotify event` following `detected rotation of /var/log/server.log; waiting 5 seconds`.  The message seems to be reported by https://github.com/fluent/fluentd/blob/v1.13.3/lib/fluent/plugin/in_tail.rb#L482. It seems to be strongly related position file having two entries.

We searched some other debug logs related this issue but we had no such logs. fluentd reported no other debug logs but only pairs of `detected rotation of /var/log/server.log; waiting 5 seconds` and `The following tail of /var/log/server.log`.

**Comment by ashie** [comment]
Thanks for your report!

> And, fluentd reported `Skip update_watcher because watcher has been already updated by other inotify event` following `detected rotation of /var/log/server.log; waiting 5 seconds`.  The message seems to be reported by https://github.com/fluent/fluentd/blob/v1.13.3/lib/fluent/plugin/in_tail.rb#L482. It seems to be strongly related position file having two entries.

To tell the truth, I was suspecting it but I couldn't confirm it because I can't yet reproduce it.
Your report it very helpful for me.

Although I'm not yet sure the mechanism of this issue, you might be able to avoid this issue by disabling stat watcher (`enable_stat_watcher false`).

**Comment by ashie** [comment]
Similar reports:

issue | date | user | fluentd version | read_from_head | follow_inodes |  work around
-- | -- | -- | -- | -- | -- | --
#3586 | 2021-12-28 | @jasminehung | v1.14.3 | true | true | `pos_file_compaction_interval 3m`
#3614 | 2022-02-02 | @salavessa | v1.14.4 | true | false |  
  | 2022-02-14 | @mizunashi-mana | 1.14.3, v1.13.3 | false | false |  
#3661 | 2022-03-08 | @shenmuxiaosen | v1.13.3 | false | false | v1.12.4 was stable
  | 2022-07-08 | @ljuaneda | ? | ? | ? | v1.11.5 was stable
#3800 | 2022-06-27 | @bjg2 | v1.13.3 | true | false |

**Comment by ashie** [comment]
There are several similar reports and most of them said:

When tailing is stopped, 
```
202x–xx–xx xx:xx:xx +0000 [info]: #0 detected rotation of /var/log/containersxxx.log; waiting 5 seconds
```
was observed but
```
202x–xx–xx xx:xx:xx +0000 [info]: #0 following tail of /var/log/containers/xxx.log
```
wasn't observed.

In addition, a user reports that following debug message is observed when log level is debug:
```
202x–xx–xx xx:xx:xx +0000 [debug]: #0 Skip update_watcher because watcher has been already updated by other inotify event
```

So, it seems caused by the following code:
https://github.com/fluent/fluentd/blob/438a82aead488a86180cd484bbc4e7e344a9032b/lib/fluent/plugin/in_tail.rb#L489-L498

**Comment by wangchaoforever** [comment]
@ashie Is this one fixed?

**Comment by ashie** [comment]
At v1.15.1 we added a log message like the following to investigate this issue:
```
2022-08-01 17:57:19 +0900 [warn]: #0 Skip update_watcher because watcher has been already updated by other inotify event path="logs/hige.log" inode=59511737 inode_in_pos_file=0
```
If you see such log message, please report it.

**Comment by vparfonov** [comment]
Hello, @ashie! We have something similar to: "no more log pushed" and found duplication in `pos` file. 
Like this:
```
/var/log/pods/pod-l4wrv/service/0.log	ffffffffffffffff	000000008b00ad36

/var/log/pods/pod-5bmg8/service/0.log	000000000333fd2e	000000005a00673b
/var/log/pods/pod-5bmg8/service/0.log	0000000004b2a353	000000005a00673e
/var/log/pods/pod-5bmg8/service/0.log	0000000003666a64	000000005a00673c
/var/log/pods/pod-5bmg8/service/0.log	0000000006a53eab	000000005a00673e
/var/log/pods/pod-5bmg8/service/0.log	0000000004a21b23	000000005a01e248
/var/log/pods/pod-5bmg8/service/0.log	0000000000000000	0000000000000000
```
Maybe this can help you.

**Comment by ashie** [comment]
@salavessa Could you provide your in_tail config? Do you use `follow_inode true`?

**Comment by vparfonov** [comment]
@ashie, looks like not, we use default setting for `follow_inode`
```
<source>
  @type tail
  @id container-input
  path "/var/log/pods/**/*.log"
  exclude_path ["/var/log/pods/openshift-logging_collector-*/*/*.log", "/var/log/pods/openshift-logging_elasticsearch-*/*/*.log", "/var/log/pods/openshift-logging_kibana-*/*/*.log"]
  pos_file "/var/lib/fluentd/pos/es-containers.log.pos"
  refresh_interval 5
  rotate_wait 5
  tag kubernetes.*
  read_from_head "true"
  skip_refresh_on_startup true
  @label @MEASURE
  <parse>
    @type multi_format
    <pattern>
      format json
      time_format '%Y-%m-%dT%H:%M:%S.%N%Z'
      keep_time_key true
    </pattern>
    <pattern>
      format regexp
      expression /^(?<time>[^\s]+) (?<stream>stdout|stderr)( (?<logtag>.))? (?<log>.*)$/
      time_format '%Y-%m-%dT%H:%M:%S.%N%:z'
      keep_time_key true
    </pattern>
  </parse>
</source>
```

**Comment by artheus** [comment]
I looked a bit at this now. What seems to happen is that in the `in_tail.rb` it checks if `File.readable?` which is false in the case of a broken symlink.
So in the case of log rotation, the symlink to the container log file will be broken for a while, in Kubernetes environments.
If there are no log lines output within the timeout of the detected log rotation, the in_tail will assume the file to be readable, but no file is actually created yet.
This could easily be fixed by polling the symlink, until it's no longer broken (or until some kind of max_retry is hit)
I also think it should be announced somewhere that one should install `capng` to get the capabilities lib working properly. Which it was not, for us. And we're using the official docker images for running fluentd

**Comment by artheus** [comment]
It is really weird however, that the re-examination done on the next refresh does not always pick up on that the symlinks are no longer broken, and that the file should now be readable.
I have however found that in most of the cases where we will get no more logs, after a detected log rotation, the "detected rotation of ..." log entry was sent twice for the same file, on the same node, and in other cases when FluentD is able to pick up tailing the file again later, the log line appears only once.

**Comment by artheus** [comment]
I created a test case in `test/plugin/test_in_tail.rb` for just playing around with this exact event:
```ruby
  sub_test_case "multiple log rotations" do
    data("fleep floop hurdy gurdy" => ["1.log", 5120, "msg"])

    def test_reproduce_err_after_rotations(data)
      file, num_lines, msg = data

      File.open("#{@tmp_dir}/#{file}", 'wb') do |f|
        num_lines.times do
          f.puts "#{msg}\n"
        end
      end

      path_element = create_path_element("*.log.link")

      conf = ROOT_CONFIG + path_element + CONFIG_READ_FROM_HEAD + SINGLE_LINE_CONFIG

      File.symlink(file, "#{@tmp_dir}/log_first.log.link")

      dl_opts = {log_level: Fluent::Log::LEVEL_DEBUG}
      logdev = $stdout
      logger = ServerEngine::DaemonLogger.new(logdev, dl_opts)
      log_instance = Fluent::Log.new(logger)

      d = create_driver(conf, false)
      d.instance.log = log_instance
      d.run(timeout: 30) do
        sleep 1

        File.rename("#{@tmp_dir}/#{file}", "#{@tmp_dir}/2.log")

        puts "Error events: #{d.error_events}"
        puts "Record count: #{d.record_count}"
        puts "Emit count: #{d.emit_count}"

        sleep 5

        File.open("#{@tmp_dir}/#{file}", 'wb') do |f|
          num_lines.times do
            f.puts "#{msg}\n"
          end
        end

        sleep 5

        puts "Error events: #{d.error_events}"
        puts "Record count: #{d.record_count}"
        puts "Emit count: #{d.emit_count}"
      end
    end
  end
```

**Comment by SriramDuvvuri** [comment]
HI All,

We are facing a similar kind of issue where the td-agent.log & .pos files are not updating properly. 
In some cases , failing after when log rotation occurs, file gets created but nothing is generated as shown below. Even though if we restart the td-agent, the problem is not fixed

As we have systems over RHEL 7/6 we are using **fluentd version using 1.11.5** td-agent 3.8.1/3.8.0, in our environment
Here is the config we are using, please provide us your inputs here.

[root]:/var/log/td-agent> ls -ltr
-rw-r--r-- 1 td-agent td-agent       68 Aug 29 20:00 s2.pos
-rw-r--r-- 1 td-agent td-agent      384 Aug 29 20:00 b1.pos
-rw-r--r-- 1 td-agent td-agent    37509 Aug 29 20:00 a3.pos
-rw-r--r-- 1 td-agent td-agent    41629 Aug 29 20:00 xx.pos
-rw-r----- 1 td-agent td-agent 18562883 Aug 29 20:38 **td-agent.log-20220830**
-rw-r----- 1 td-agent td-agent        **0 Aug 30 03:50 td-agent.log**
****

Config:
---------
<source>
@type tail
  path /home/server/xx*.log
  exclude_path ["/home/server/xxyy*.log"]
  pos_file /var/log/td-agent/xxlog.pos
  tag xx.log
  enable_stat_watcher false
  read_from_head true
  follow_inodes true
  refresh_interval 5
  rotate_wait 5
  skip_refresh_on_startup true
  pos_file_compaction_interval 30m
  <parse>
    @type multiline
    format_firstline /<20/
    format1 /<(?<logDate>.*?)>-<(?<logThread>.*?)>-<(?<level>.*?)>-<(?<logClass>.*?)>-(?<logMessage>.*)/
    time_key logDate
    time_format %Y/%m/%d-%H:%M:%S.%N
  </parse>
</source>

PS. the same config is working fine in few other instances. I am not sure what was the problem here.

Appreciate your help in advance.

**Comment by artheus** [comment]
We just ran into the same problem again. Without the log line `Skip update_watcher because watcher has been....` showing up.
Now we only get these:
```
2022-09-13 06:36:37 +0000 [warn]: #0 [in_tail_container_logs] got incomplete line at shutdown from /var/log/containers/redacted-service-name-6b6754b75c-swn6z_backend_redacted-service-name-13f36bd8f73334e603ee6fe80642d971f6ea0b25ce966c6ec494c5b226e4d93c.log: ""
```

Is it wrong that we are running this in a k8s DaemonSet? Should we install FluentD/FluentBit directly on the vm's instead?
Are there any special capabilities needed for the Pods created by the DaemonSet?
I'm starting to feel a bit lost in this problem...

**Comment by artheus** [comment]
Are there any workarounds for this, that anyone has been able to detect?

My current thoughts are:
* Run FluentD on VM directly, instead of running it as a DaemonSet in K8S (w. mounted /var/log)
* Create our own tail plugin, that might be able to to a better job with tailing and log rotations.
  * E.g. Not only watch the symlinks, but also watch the files symlinks point to, to be able to get an event-driven async way of detecting log rotation, and when new log files are created.
* Alternatively create some middle-man app that will tail and detect log rotations properly, and send logs to FluentD via e.g. http

Please contribute with whatever solutions you've set up to get this working as expected.
In our company, we must follow government laws about audit logging. So we cannot afford to loose logs due to this bug.

**Comment by artheus** [comment]
> > We just ran into the same problem again. Without the log line `Skip update_watcher because watcher has been....` showing up. Now we only get these:
> 
> Are you using fluentd v1.15.1 or later?

Yes. We are currently using the Docker image [fluent/fluentd:v1.15.1-debian-1.1](https://hub.docker.com/layers/fluent/fluentd/v1.15.1-debian-armhf-1.1/images/sha256-7b4b08befd59ebbb6013076869253bee48afca2155dc9254f45d39ba7e49356d?context=explore)

**Comment by zmx** [comment]
Fluentd v1.15.1 with `enable_stat_watcher false` and `pos_file_compaction_interval 3m` and `follow_inodes true`  still drop 50~70% logs for us. The load is around 20k logs/s

We see the pos file have duplicate entries and all zero inoder number

Our workaround: Using fluentbit in_tail plugin, then out_forward to fluentd using existing rules.

**Comment by zmx** [comment]
This bug happens on our environment for version 1.15.1 with all workaround
above. Drop almost 70% logs with 20k logs/s.

We see exactly the same POS  file corrupt mention in this ticket.

Our workaround: change to fluentbit tail plugin , out_forward plugin to
fluentd to keep existing rules.

Morten Hekkvang ***@***.***>於 2022年10月10日 週一，下午3:55寫道：

**Comment by masterashu-motorq** [comment]
> At v1.15.1 we added a log message like the following to investigate this issue:
> 
> ```
> 2022-08-01 17:57:19 +0900 [warn]: #0 Skip update_watcher because watcher has been already updated by other inotify event path="logs/hige.log" inode=59511737 inode_in_pos_file=0
> ```
> 
> If you see such log message, please report it.

Getting similar log
```log
2022-11-14 23:14:57 +0000 [info]: #0 detected rotation of /var/log/pods/processor-0-65dc565cc8-hq469_0f06669b-4521-4297-80db-ead8aaecde4e/core-ford-processor-0/1.log; waiting 5 seconds
2022-11-14 23:14:57 +0000 [warn]: #0 Skip update_watcher because watcher has been already updated by other inotify event path="/var/log/pods/processor-0-65dc565cc8-hq469_0f06669b-4521-4297-80db-ead8aaecde4e/core-ford-processor-0/1.log" inode=4397239 inode_in_pos_file=0
```

**Comment by hawk-bernhard-altendorfer** [comment]
fluentd v1.15-debian-opensearch-1 -> gem 'fluentd' version '1.15.3'

````
<source>
              @type tail
              @id in_tail_container_logs
              path /var/log/containers/*.log
              follow_inodes true
              refresh_interval 10
              rotate_wait 1
              enable_stat_watcher false
              pos_file /var/log/fluentd-containers.log.pos
              pos_file_compaction_interval 12h
              tag raw.kubernetes.*
              read_from_head true
              read_bytes_limit_per_second 1048576000
              max_line_size 2MB
              <parse>
                @type multi_format
                <pattern>
                  format json
                  time_key time
                  time_type string
                  time_format "%Y-%m-%dT%H:%M:%S.%NZ"
                  keep_time_key false
                </pattern>
                <pattern>
                  format regexp
                  expression /^(?<time>.+) (?<stream>stdout|stderr)( (.))? (?<log>.*)$/
                  time_format '%Y-%m-%dT%H:%M:%S.%NZ'
                  keep_time_key false
                </pattern>
              </parse>
              emit_unmatched_lines true
</source>
```

Last logs of fluentd before pushing stops:

```
2023-01-25 17:55:53 +0000 [info]: #0 [in_tail_container_logs] detected rotation of /var/log/containers/gateway-rest-68f6cb4bb4-vlfcq_default_gateway-rest-23c1ac09792e85465483b2d641e58fa6d5d8181924bc7b08f739779b2bf1d02f.log; waiting 1.0 seconds

2023-01-25 17:55:53 +0000 [info]: #0 [in_tail_container_logs] following tail of /var/log/containers/gateway-rest-68f6cb4bb4-vlfcq_default_gateway-rest-23c1ac09792e85465483b2d641e58fa6d5d8181924bc7b08f739779b2bf1d02f.log
```

When I restart fluentd then it starts to push all the logs that were missing.

**Comment by kevinsookocheff-wf** [comment]
We encountered the same error. I've been trying to reproduce without success so far. In our case we did a large upgrade from version 1.6.3 to 1.15.3.

**Comment by jcantrill** [comment]
@ashie we have a customer who claims they applied the following patch and the issue goes away.  I do not have a reliable way to reproduce the problem.  We are using v1.14.6

```--- in_tail.rb.org      2023-02-13 17:04:42.353997075 +0900
+++ in_tail.rb  2023-02-13 17:04:39.030016491 +0900
@@ -354,13 +354,15 @@

     def existence_path
       hash = {}
-      @tails.each_key {|target_info|
-        if @follow_inodes
-          hash[target_info.ino] = target_info
-        else
-          hash[target_info.path] = target_info
-        end
-      }
+      if @follow_inodes
+        @tails.each {|ino, tw|
+          hash[tw.ino] = TargetInfo.new(tw.path, tw.ino)
+        }
+      else
+        @tails.each {|path, tw|
+          hash[tw.path] = TargetInfo.new(tw.path, tw.ino)
+        }
+      end
       hash
     end

@@ -441,8 +443,11 @@

       begin
         target_info = TargetInfo.new(target_info.path, Fluent::FileWrapper.stat(target_info.path).ino)
-        @tails.delete(target_info)
-        @tails[target_info] = tw
+        if @follow_inodes
+          @tails[target_info.ino] = tw
+        else
+          @tails[target_info.path] = tw
+        end
         tw.on_notify
       rescue Errno::ENOENT, Errno::EACCES => e
         $log.warn "stat() for #{target_info.path} failed with #{e.class.name}. Drop tail watcher for now."
@@ -462,9 +467,17 @@
     def stop_watchers(targets_info, immediate: false, unwatched: false, remove_watcher: true)
       targets_info.each_value { |target_info|
         if remove_watcher
-          tw = @tails.delete(target_info)
+          if @follow_inodes
+            tw = @tails.delete(target_info.ino)
+          else
+            tw = @tails.delete(target_info.path)
+          end
         else
-          tw = @tails[target_info]
+          if @follow_inodes
+            tw = @tails[target_info.ino]
+          else
+            tw = @tails[target_info.path]
+          end
         end
         if tw
           tw.unwatched = unwatched
@@ -478,10 +491,19 @@
     end

     def close_watcher_handles
-      @tails.keys.each do |target_info|
-        tw = @tails.delete(target_info)
-        if tw
-          tw.close
+      if @follow_inodes
+        @tails.keys.each do |ino|
+          tw = @tails.delete(ino)
+          if tw
+            tw.close
+          end
+        end
+      else
+        @tails.keys.each do |path|
+          tw = @tails.delete(path)
+          if tw
+            tw.close
+          end
         end
       end
     end
@@ -500,26 +522,25 @@
       end

       rotated_target_info = TargetInfo.new(target_info.path, pe.read_inode)
-      rotated_tw = @tails[rotated_target_info]
-      new_target_info = target_info.dup

       if @follow_inodes
+        rotated_tw = @tails[rotated_target_info.ino]
+        new_target_info = target_info.dup
         new_position_entry = @pf[target_info]

         if new_position_entry.read_inode == 0
           # When follow_inodes is true, it's not cleaned up by refresh_watcher.
           # So it should be unwatched here explicitly.
           rotated_tw.unwatched = true
-          # Make sure to delete old key, it has a different ino while the hash key is same.
-          @tails.delete(rotated_target_info)
-          @tails[new_target_info] = setup_watcher(new_target_info, new_position_entry)
-          @tails[new_target_info].on_notify
+          @tails[new_target_info.ino] = setup_watcher(new_target_info, new_position_entry)
+          @tails[new_target_info.ino].on_notify
         end
       else
-        # Make sure to delete old key, it has a different ino while the hash key is same.
-        @tails.delete(rotated_target_info)
-        @tails[new_target_info] = setup_watcher(new_target_info, pe)
-        @tails[new_target_info].on_notify
+        rotated_tw = @tails[rotated_target_info.path]
+        new_target_info = target_info.dup
+
+        @tails[new_target_info.path] = setup_watcher(new_target_info, pe)
+        @tails[new_target_info.path].on_notify
       end
       detach_watcher_after_rotate_wait(rotated_tw, pe.read_inode) if rotated_tw
     end

--- position_file.rb.org        2023-02-13 15:02:14.671608849 +0900
+++ position_file.rb    2023-02-13 15:06:09.098223852 +0900
@@ -250,20 +250,6 @@
       end
     end

-    TargetInfo = Struct.new(:path, :ino) do
-      def ==(other)
-        return false unless other.is_a?(TargetInfo)
-        self.path == other.path
-      end
-
-      def hash
-        self.path.hash
-      end
-
-      def eql?(other)
-        return false unless other.is_a?(TargetInfo)
-        self.path == other.path
-      end
-    end
+    TargetInfo = Struct.new(:path, :ino)
   end
 end
```

**Comment by jcantrill** [comment]
@ashie I believe the issue is because the filesystem reuse of inodes and fluent using inode as the "key" for maintaining watches.  Our customer provided the following where then position entry is converted to decimal:
```
-- Pos file
/var/log/pods/openshift-image-registry_node-ca-k6thz_78bc195f-591c-4af5-afb1-a31c601850e9/node-ca/0.log 18446744073709551616    195064732
/var/log/pods/openshift-image-registry_node-ca-k6thz_78bc195f-591c-4af5-afb1-a31c601850e9/node-ca/0.log 18446744073709551616    195064733
/var/log/pods/openshift-image-registry_node-ca-k6thz_78bc195f-591c-4af5-afb1-a31c601850e9/node-ca/0.log 58444023        195064732
/var/log/pods/openshift-image-registry_node-ca-k6thz_78bc195f-591c-4af5-afb1-a31c601850e9/node-ca/0.log 35228450        195064763
```
Note there are multiple entries with the same inode where one is identified as "unwatched" and the other is still open.  The logic then closes the watch on the inode even though it is still in use

**Comment by ashie** [comment]
@jcantrill Thanks for your information! We'll check it.

**Comment by parandiary** [comment]
td-agent 4.4.2 fluentd 1.15.3 버전을 사용하고 있습니다.
`td-agent --version
td-agent 4.4.2 fluentd 1.15.3 (e89092ce1132a933c12bb23fe8c9323c07ca81f5)
`

I'm tailing logs by mounting EFS on AWS EC2, and I can't track the tail after the log file is rotated.
Here's the log when tracking fails:
`
--- success watch
2023-03-17 00:00:02 +0900 [info]: #0 [input_tail] following tail of /EFS/STG/app_log/app1/ims.log
--- fail watch
2023-03-18 00:00:02 +0900 [info]: #0 [input_tail] following tail of /EFS/STG/app_log/app1/ims.log
2023-03-18 00:00:02 +0900 [info]: #0 [input_tail] detected rotation of /EFS/STG/app_log/app1/ims.log; waiting 5 seconds
--- success watch
2023-03-19 00:00:02 +0900 [info]: #0 [input_tail] following tail of /EFS/STG/app_log/app1/ims.log
--- fail watch
2023-03-20 00:00:02 +0900 [info]: #0 [input_tail] following tail of /EFS/STG/app_log/app1/ims.log
2023-03-20 00:00:02 +0900 [info]: #0 [input_tail] detected rotation of /EFS/STG/app_log/app1/ims.log; waiting 5 seconds
-- success watch
2023-03-21 00:00:02 +0900 [info]: #0 [input_tail] following tail of /EFS/STG/app_log/app1/ims.log
`


I tried with the settings below, but it failed and I got an unwatched log after 7 minutes of detect rotation.
`
  enable_stat_watcher false
  refresh_interval 3s
  rotate_wait 1s
  read_from_head true
  follow_inodes true
  pos_file_compaction_interval 10m
`

`
2023-03-22 00:00:01 +0900 [debug]: #0 [input_tail] tailing paths: target = /EFS/STG/app_log/app1/ims.log | existing = /EFS/STG/app_log/app1/ims.log
2023-03-22 00:00:01 +0900 [info]: #0 [input_tail] following tail of /EFS/STG/app_log/app1/ims.log
2023-03-22 00:00:01 +0900 [info]: #0 [input_tail] detected rotation of /EFS/STG/app_log/app1/ims.log; waiting 1 seconds
2023-03-22 00:00:16 +0900 [debug]: #0 [input_tail] tailing paths: target = /EFS/STG/app_log/app1/ims.log | existing = /EFS/STG/app_log/app1/ims.log
2023-03-22 00:07:16 +0900 [debug]: #0 [input_tail] Remove unwatched line from pos_file: /EFS/STG/app_log/app1/ims.log   ffffffffffffffff        9a58588af4b55d9f
`

If you try with the settings below, the log is tracked normally, but there are concerns about performance. Please refer to the log and advise on performance.
`
  enable_stat_watcher false
  refresh_interval 3s
  rotate_wait 1s
  read_from_head true
  follow_inodes true
  pos_file_compaction_interval 3m
  **open_on_every_update true**
`

log
`
2023-03-21 23:58:07 +0900 [info]: #0 [input_tail] Clean up the pos file
2023-03-22 00:00:41 +0900 [info]: #0 [input_tail] detected rotation of /EFS/PRD/app_log/app1/ims.log
2023-03-22 00:00:41 +0900 [info]: #0 [input_tail] following tail of /EFS/PRD/app_log/app1/ims.log
2023-03-22 00:00:44 +0900 [info]: #0 [input_tail] following tail of /EFS/PRD/app_log/app1/ims.log
2023-03-22 00:01:07 +0900 [info]: #0 [input_tail] Clean up the pos file
`

**Comment by jcantrill** [comment]
@ashie I have done some rudimentary investigation using the following:

* conf: https://gist.github.com/jcantrill/99d913f0def1b87719b402078e258a4c
* harness: https://gist.github.com/jcantrill/4dafdf19ef3acea1e716fb4fdb787e9d

I set up 4 separate terminals:

* one terminal to run the test harness
* one fluentd `fluentd -c position.conf --no-supervisor`
* one watching the count of the position file
* one watching the file handles `lsof -nPw -p $pid | grep /loopfs | grep log | wc -l"`

The basic testing I have done keeps the writer size to 512 and various the number of writers from 1-250.  My summary of what I observed is in line with what I would expect:  as you increase the files being tailed, fluent becomes CPU bound and is not able to predictably keep up with the load.  In a containerized environment like kubernetes, it will miss full log rotations and segments of logs.  Changing the position file compact interval from the default of 72h to 45s helps but it in no way can account for "missing watches".  

Our customer claims fluend is closing these watches because of multiple entries in the position file but I do not believe that to be the case.  It seems to me that we are simply running up against the limitations of ruby and fluent and the fact it is not able to efficiently process tailing a large number of files that are constantly being rotated.  I do see evidence of times where there are "missing watches"; the number is less then the number of files producing logs but these are few and it eventually recovers.

I'm not convinced there is an issue other then hitting the performance limitations of fluentd

**Comment by kattz-kawa** [comment]
> we have a customer who claims they applied the following patch and the issue goes away. I do not have a reliable way to reproduce the problem. We are using v1.14.6

This is my proposed patch for the customer.

@jcantrill 
Thank you for validating this patch.

> I'm going to walk back my statement to some after applying the proposed patch and doing similar rudimentary testing. The patch demonstrates an improvement with "missed" file rotations. I can not speak, however, to whether:

@ashie

This issue is observed by Jeff's primitive test code.
Additionally this issue also is observed on the popular enviroment like as using 'logrotate' program.
(such as k8s)

I attach the simple reproducer program using 'logrotate'.

the reproducible way is below:

<pre>
Test environment:
  RHEL8.7 on KVM(2 vCPUs, 4GB memory)

Kernel version:
  Linux 4.18.0-425.3.1.el8.x86_64

I tested it on RHEL8.7(kernel-4.18.0-425.3.1.el8.x86_64) KVM, 
However this procedure could run even on another Linux since it runs as a container.
Please replace all "podman" string to "docker" in test_run.sh , this include in reproducer.tar.gz, 
if your linux has no podman command.

Steps to reproduce:

1. Untar 'reproducer.tar.gz' on your machine

# tar zxf reproducer.tar.gz

2. Run 'test_run.sh' script

# nohup bash test_run.sh &

3. Wait until the "The issue has happend." message is printed in the 'lrt.log'
　 'lrt.log' will be generated on the current directory.
　 Run 'tail -f' command to monitor that "The issue has happend" message is printed on 'lrt.log'.

    Ex) The issue has happened. Mon Mar 13 07:23:16 JST 2023

4. Confirm that fluentd has already closed '/var/log/pods/test/0.log' although the file still exists.

     1) Check INODE of '/var/log/pods/test/0.log'

        # podman exec -it reproducer /bin/bash
        # ls -li /var/log/pods/test/0.log

     2) Check Fluentd never track the '/var/log/pods/test/0.log'

        # lsof -p <fluentd pid> | grep <INODE of 0.log file>
          => It's closed although Fluentd has to collect the '/var/log/pods/test/0.log'

     3) Check the position file if the '/var/log/pods/test/0.log' should be tracked

        # awk "{printf(\"%s\t%s\t%s\n\", \$1, strtonum(\"0x\"\$2), strtonum(\"0x\"\$3))}"  /var/lib/fluentd/pos/es-containers.log.pos | grep <INODE of 0.log file>
          => If the inode exists, it means the log file should be collected by Fluentd

        ex)
           /var/log/pods/test/0.log        20165   486593998

     4) Check if the '/var/log/pods/test/0.log' is not tracked any more
  
        # kill -s SIGCONT {logger.sh pid}
        => Resume to send test logs to the '/var/log/pods/test/0.log'

        However Fluent won't reopen/reread the '/var/log/pods/test/0.log'
        As far as I can see this issue happen if 'follow_inodes' set 'true'.
</pre>

I will try to create a pull request on this issue later.
Could you review it?
I hope this attached program is helpful to you to validate my patch.
[reproducer.tar.gz](https://github.com/fluent/fluentd/files/11485631/reproducer.tar.gz)

**Comment by ashie** [comment]
Thanks for your comment, and sorry for my late response.
Although I think this issue should have high priority but I'm not yet able to check the detail because I'm bothered by dozens of trivial tasks...

> I will try to create a pull request on this issue later.
> Could you review it?

Of course pull requests are welcome!
In addition, discussing about the code here is hard, using a pull request is recommended for it.

>> we have a customer who claims they applied the following patch and the issue goes away. I do not have a reliable way to reproduce the problem. We are using v1.14.6
>
> This is my proposed patch for the customer.

Interesting, but I have some concerns about it.

* As in https://github.com/fluent/fluentd/issues/3614#issuecomment-1179049132 originally this issue treated  mainly `follow_inode false` case, but your patch seem effects only to `follow_inode true` case.
* Using inode as keys might have side effect, we might experience such case in the past.
  * I don't yet remember the detail. Probably one of side effect is log duplication.
* It's not based on latest code.

**Comment by ashie** [comment]
>     * Using inode as keys might have side effect, we might experience such case in the past.
>       * I don't yet remember the detail. Probably one of side effect is log duplication.

On the other hand, certainly I also think that it is strange to manage with path in `follow_inode true` case.
In addition, probably the case we experienced before is mixed other conditions. I'm digging out it.

**Comment by kattz-kawa** [comment]
> #4190 describes a cause of in_tails's stall issue very clearly.
> It's indeed one of major cause of this issue.
> I'm not sure whether other causes are still remained or not though.

Thank you for the information!
It looks like that we no longer face the issue with my reproduction script after applying the patch of #4190.

> * Using inode as keys might have side effect, we might experience such case in the past.
>  * I don't yet remember the detail. Probably one of side effect is log duplication.

Our approach is based on below:
- The 'tails' hash key which using path will be collision when inode is reused.
- We suggest that the 'tails' should use 'inode' key in order not to cause the simillar issue which inode reuse case to the feature release.

But We also can agree that some unexpected side effect might be caused by that since our patch would be the big change compared with #4190.

Our only concern for #4190 is if the patch might cause a new file handle leak issue.
We don't know really if there is no case that `tw.ino != ino` can be true except for #4190 case.
(Although I didn't face such a leak issue during the test on my local env)

**Comment by k-keiichi-rh** [comment]
>> On the other hand, certainly I also think that it is strange to manage with path in follow_inode true case. In addition, probably the case we experienced before is mixed other conditions. I'm digging out it.
>
> strongly +1. Besides, regarding to code structure, the in_tail code has lots of "if follow_inode ... else ..." logic, which makes the code hard to read.

I have no objection this issues is should be fixed as described in https://github.com/fluent/fluentd/issues/4190.

But mixing inode and path to manage the hashes makes us confused when follow_inodes is true. It seems like it's indicated by this long discussion in this ticket from early 2022.
So I think this situation should be resolved as @ashie and @garyzjq commented. It may need a code refactoring for the "if follow_inode ... else ..." logic. But it will be another topic...

**Comment by ashie** [comment]
The reason why I didn't apply the patch in https://github.com/fluent/fluentd/issues/3614#issuecomment-1439046528 soon is that:

* There was no description about the detailed mechanism of the issue.
* So I couldn't understand why this patch fix the issue.

But when I read #4190, the question has been resolved.
Yes, #4185 certainly resolves the issue because rotated inodes aren't included in `@tails`.

Now I'm basically positive to merging #4190, I'll recheck it in some perspective though.
The main issue is that I want to add tests for it (but it might be too hard).

On the other hand, I'm also considering to merge #4191 mainly for `follow_inodes false` case.
It seems that both of them can coexist.

> Our only concern for https://github.com/fluent/fluentd/issues/4190 is if the patch might cause a new file handle leak issue.

In my understanding, it doesn't leak handles, they will be scheduled to free by rotate_watcher in this case.

Anyway, thanks to you guys, we've found the road to solution.
Thank you so much!

**Comment by daipom** [comment]
Thanks to the fix of #4185 and the description of #4190, I understood this problem as follows.

Before rotate

*  container.log (inodeA)

After rotate

* container.log (inodeB)
* container.backup.log (inodeA)

https://github.com/fluent/fluentd/blob/dcc05e59e038ffef16ee2b510d3d9e34b23bbdfe/lib/fluent/plugin/in_tail.rb#L370-L386

The `added_hash` will be `{inodeB => TargetInfo("container.log", inodeB)}`.
So, `construct_watcher` will be called to construct `TargetInfo("container.log", inodeB)`.

https://github.com/fluent/fluentd/blob/dcc05e59e038ffef16ee2b510d3d9e34b23bbdfe/lib/fluent/plugin/in_tail.rb#L422-L448

This will overwrite `@tails["container.log"]` for `container.log (inodeB)`.
However, `@tail["container.log"]` was originally used for `container.log (inodeA)`.
The old `TailWatcher` (for `container.log (inodeA)`) can call `update_watcher` to detach `@tail["container.log"]`.
This will result in the new rotate file being wrongly detached.

https://github.com/fluent/fluentd/blob/dcc05e59e038ffef16ee2b510d3d9e34b23bbdfe/lib/fluent/plugin/in_tail.rb#L502-L519

**Comment by daipom** [comment]
The problem with this scenario is that the keys of `@tails` are the paths even if `following_inode` is enabled.
I too think the fix of #4185 will solve the problem in this scenario.

I would like to see more details.

**Comment by daipom** [comment]
About what we are talking about in #4185.
I think the problem with the above scenario (https://github.com/fluent/fluentd/issues/3614#issuecomment-1578977365) is that `refreash_watcher` is trying to update the paths of the existing `TailWatcher`s.
This causes conflicts.

The existing `TailWatcher`s will be updated by `StatWatcher` or `TimerTrigger`, so `refresh_watcher` shouldn't update them.
(Is this correct?)

However, when `follow_inodes` is enabled, `refresh_watcher` determines the differences only in the inodes.
It needs to also see `path` because the existing `TailWatcher`s are updated based on its' paths, even when `following_inodes` is enabled.

For example, I think the following modifications may provide a partial solution.

```diff
diff --git a/lib/fluent/plugin/in_tail.rb b/lib/fluent/plugin/in_tail.rb
index fb29dd24..20b6cd4f 100644
--- a/lib/fluent/plugin/in_tail.rb
+++ b/lib/fluent/plugin/in_tail.rb
@@ -380,6 +380,11 @@ module Fluent::Plugin
       unwatched_hash = existence_paths_hash.reject {|key, value| target_paths_hash.key?(key)}
       added_hash = target_paths_hash.reject {|key, value| existence_paths_hash.key?(key)}
 
+      if @follow_inodes
+        unwatched_hash.reject! {|ino, target_info| existence_paths_hash.value?(target_info.path)}
+        added_hash.reject! {|ino, target_info| existence_paths_hash.value?(target_info.path)}
+      end
+
       stop_watchers(unwatched_hash, immediate: false, unwatched: true) unless unwatched_hash.empty?
       start_watchers(added_hash) unless added_hash.empty?
       @startup = false if @startup
```

**Comment by kattz-kawa** [comment]
@daipom 

Thank you for suggesting another solution!

Just a quick reporting on my reproducer with applying your suggested patch,
the result is not good. This issue happened.
I'm digging the reason why this issue happened with your patch.

**Comment by daipom** [comment]
@kattz-kawa @masaki-hatada Thanks so much for confirming the behavior!
I'm sorry. I did an easy mistake.
The patch I was thinking of is as follows.
However, the tests don't succeed in this patch.
I will reconsider this direction and this patch.

```diff
diff --git a/lib/fluent/plugin/in_tail.rb b/lib/fluent/plugin/in_tail.rb
index fb29dd24..d13407f0 100644
--- a/lib/fluent/plugin/in_tail.rb
+++ b/lib/fluent/plugin/in_tail.rb
@@ -380,6 +380,12 @@ module Fluent::Plugin
       unwatched_hash = existence_paths_hash.reject {|key, value| target_paths_hash.key?(key)}
       added_hash = target_paths_hash.reject {|key, value| existence_paths_hash.key?(key)}
 
+      if @follow_inodes
+        paths = existence_paths_hash.collect {|ino, target_info| target_info.path}
+        unwatched_hash.reject! {|ino, target_info| paths.any?(target_info.path)}
+        added_hash.reject! {|ino, target_info| paths.any?(target_info.path)}
+      end
+
       stop_watchers(unwatched_hash, immediate: false, unwatched: true) unless unwatched_hash.empty?
       start_watchers(added_hash) unless added_hash.empty?
       @startup = false if @startup
```

**Comment by kattz-kawa** [comment]
Hi @daipom

Thank you for deeply discussing this issue.
I'm testing the patch in [#3614(comment)](https://github.com/fluent/fluentd/issues/3614#issuecomment-1580278892) last week.

It looks good for unexpected file close issue 
But the 'tw' leak issue probably occurred you know.

<pre>
pid 413 = fluentd process

# lsof -p 413 -nP | grep "/var/log/pods/" | grep deleted
ruby-mri 413 root   10r      REG   0,64    24198 167792839 /var/log/pods/test/0.log.8 (deleted)
ruby-mri 413 root   22r      REG   0,64    20660 360740488 /var/log/pods/dummy102/0.log.8 (deleted)
</pre>

You may have already noticed, but I leave this information in the comments just in case.

**Comment by daipom** [comment]
@kattz-kawa Thanks so much! I didn't notice it!
Sorry for my lack of progress.
I want to make some conclusion this week.

**Comment by daipom** [comment]
I think the following is one of the causes.

https://github.com/fluent/fluentd/blob/0a6d706a9cee5882d751b2cc6169696709df0134/lib/fluent/plugin/in_tail.rb#L504-L518

In `update_watcher`, we don't update a watcher in the case of `if new_position_entry.read_inode != 0`.
However, in this case, `detach_watcher_after_rotate_wait` is still called.

This seems wrong.
We must not detach a watcher without removing it.
(It is only allowed on closing.)

We need to fix it as follows.

```diff
           rotated_tw.unwatched = true if rotated_tw
           @tails[path] = setup_watcher(target_info, new_position_entry)
           @tails[path].on_notify
+          detach_watcher_after_rotate_wait(rotated_tw, pe.read_inode) if rotated_tw
         end
       else
         @tails[path] = setup_watcher(target_info, pe)
         @tails[path].on_notify
+        detach_watcher_after_rotate_wait(rotated_tw, pe.read_inode) if rotated_tw
       end
-      detach_watcher_after_rotate_wait(rotated_tw, pe.read_inode) if rotated_tw
```

I don't understand what the condition `new_position_entry.read_inode == 0` means. 
Certainly, this condition protects against the case where `refresh_watchers` handles the rotation first.
However, I'm not sure if the conditions were intended for it.
What is certain is that the fact that `new_position_entry.read_inode` is NOT zero here means that the inode after rotation is NOT a new inode, and we must not detach the watcher in this case.

**Comment by daipom** [comment]
I'm creating some test codes for this issue.
Based on the tests, I'd like to figure out how to adapt the modifications that have come up so far.

**Comment by daipom** [comment]
https://github.com/fluent/fluentd/issues/3614#issuecomment-1598505831

I confirmed this behavior with the #4207 tests.
This solves the problem of stopping pushing logs, but handle leaks still occur (the same as the current impl).

**Comment by masaki-hatada** [comment]
@daipom san,

Thank you for the reply!

> I think the following is one of the causes.
> 
> https://github.com/fluent/fluentd/blob/0a6d706a9cee5882d751b2cc6169696709df0134/lib/fluent/plugin/in_tail.rb#L504-L518
> 
> In `update_watcher`, we don't update a watcher in the case of `if new_position_entry.read_inode != 0`. However, in this case, `detach_watcher_after_rotate_wait` is still called.
> 
> This seems wrong. We must not detach a watcher without removing it. (It is only allowed on closing.)
> 
> We need to fix it as follows.
> 
> ```diff
>            rotated_tw.unwatched = true if rotated_tw
>            @tails[path] = setup_watcher(target_info, new_position_entry)
>            @tails[path].on_notify
> +          detach_watcher_after_rotate_wait(rotated_tw, pe.read_inode) if rotated_tw
>          end
>        else
>          @tails[path] = setup_watcher(target_info, pe)
>          @tails[path].on_notify
> +        detach_watcher_after_rotate_wait(rotated_tw, pe.read_inode) if rotated_tw
>        end
> -      detach_watcher_after_rotate_wait(rotated_tw, pe.read_inode) if rotated_tw
> ```
> 
> I don't understand what the condition `new_position_entry.read_inode == 0` means. Certainly, this condition protects against the case where `refresh_watchers` handles the rotation first. However, I'm not sure if the conditions were intended for it. What is certain is that `new_position_entry.read_inode` is NOT zero here means that the inode after rotation is NOT a new inode, and we must not detach the watcher in this case.

We also had the same concern but hesitated to change this part by the following reasons:
* This is not relating to a phenomenon reported in #3614 .
* This code was implemented 3 years ago by the following commit. We don't know why `detach_watcher_after_rotate_wait()` wasn't moved under `if new_position_entry.read_inode == 0` block in this commit.
https://github.com/fluent/fluentd/commit/b83f73e5c8830a8e56bc693ad8a6a3e316a9dcfd#diff-456fdeb51bc472beb48891caac0d063e0073655dba7ac2b72e6fdc67dc6ac802R477
* We hadn't faced any real issues about this in our test.

I agree with you that `detach_watcher_after_rotate_wait()` should be moved as you commented.

> in_tail: Ensure to detach correct watcher on rotation with follow_inodes #4208

#4208 is the replacement of #4185, isn't it?
We will check #4185.
(Thank you for adding me and @kattz-kawa as Co-author ^_^)

**Comment by daipom** [comment]
@masaki-hatada 

> https://github.com/fluent/fluentd/pull/4208 is the replacement of https://github.com/fluent/fluentd/pull/4185, isn't it?
We will check https://github.com/fluent/fluentd/pull/4185.

I would appreciate it if you could check #4208 too.
I want to merge one of #4208 and #4185. (I will review #4191 later)
Both would improve this problem.
I think #4208 would be a more direct solution, but I want to hear opinions.

> (Thank you for adding me and @kattz-kawa as Co-author ^_^)

I can create #4208 because of #4185 and #4191!
Until these PRs were created, I had no idea what was causing this problem.
Thank you all for your contributions!

**Comment by masaki-hatada** [comment]
>> We will check https://github.com/fluent/fluentd/pull/4185.
> I would appreciate it if you could check https://github.com/fluent/fluentd/pull/4208 too.

Sorry, "We will check https://github.com/fluent/fluentd/pull/4185" was a typo ^^;
I will check #4208!

**Comment by ashie** [comment]
Not yet confirmed for `follow_inode false` case. Reopen.

**Comment by ashie** [comment]
> Not yet confirmed for `follow_inode false` case. Reopen.

The mechanism of #4190 doesn't depend on `follow_inode`, so it definitely affects to `follow_inode false` case too and #4208 should fixes it. I believe #4190 is the root cause of this issue.

I'll close this issue after we check it.

**Comment by daipom** [comment]
Sorry, I have not been able to confirm `follow_inodes false` case yet. I will in the near future.

**Comment by daipom** [comment]
I added a test to reproduce `follow_inodes false` problem.

* #4264

It is not fixed yet.
We need to fix it.

I understand the mechanism.
I will explain it in detail later.

**Comment by jcantrill** [comment]
@ashie @daipom  Please be aware of a possible issue with a leak in file handles.  We have a [report](https://issues.redhat.com/browse/LOG-4241) of deleted file handles that are not being released by fluentd.  These handles are to files that have been rotated.  The issue manifests itself under high log volume in a kubernetes environment where the files are being rotated frequently and inodes are being reused.

**Comment by tmavrakis** [comment]
Any update on the case with follow_inode false? @daipom @ashie
Still getting "Skip update_watcher ..." log.

Fluentd image: fluent/fluentd:v1.16.2-1.0
Configuration: 
path "/var/log/service.log"
  path_key log_path
  refresh_interval 20 
  pos_file "/var/log/fluentd.pos"
  read_from_head true
  enable_stat_watcher false
  pos_file_compaction_interval 60s
  rotate_wait 1s

**Comment by daipom** [comment]
@tmavrakis  Thanks for your report.
I know `follow_inode false` still has a problem.
Sorry I can't make time for this problem.
I will handle this in September.

Could you share the content of the pos file when the problem occurs?

**Comment by tmavrakis** [comment]
Thanks for the update @daipom.
pos file seems to be ok.
`/var/log/service.log    000000000097d9f3        00000000002f710b`

**Comment by daipom** [comment]
> I added a test to reproduce `follow_inodes false` problem.
> 
>     * [in_tail: add test updating TailWatcher without follow_inodes #4264](https://github.com/fluent/fluentd/pull/4264)
> 
> 
> It is not fixed yet. We need to fix it.
> 
> I understand the mechanism. I will explain it in detail later.

Sorry for being late.

I summarize this `follow_inodes false` problem below.

When `follow_inodes false`, `in_tail` can not tail a new current log file if certain conditions are met.
I think this is very severe conditions.
I assume that some environments are likely to have this problem and others not so much.

The main cause is the [rotate_wait](https://docs.fluentd.org/input/tail#rotate_wait) feature.
`refresh_watcher()` can wrongly unwatch a TailWathcer still working, and then it can stop the TailWatcher's update.

Here is the mechanism.

* `refresh_watcher()` calls `stop_watcher()` when rotating occurs.
  * It is possible if `refresh_watcher()` is called when the current log file does not exist. (The previous current file is renamed, but the new file is not created yet.)
* `update_watcher()` for the rotation occurs after the `refresh_watcher()` and within `retate_wait`.
  * This updates the TailWatcher.
  * I think this is unlikely to occur in many environments. Normally, the [enable_stat_watcher](https://docs.fluentd.org/input/tail#enable_stat_watcher) feature calls `update_watcher()` immediately, before `refresh_watcher()`.
  * It may be more likely to occur in environments where the `enable_stat_watcher` feature is slow for some reason or `enable_stat_watcher` is set to `false`.
* Then, after the `rotate_wait` interval, the position is unwatched, but that position is still used in the updated TailWatcher.
  * The PositionEntry is deleted from the run-time data (`PositionFile::map`).
  * However, the TailWatcher has the PositionEntry and updates the pos file continuously.
    * Since the TailWatcher updates the PositionEntry continuously, the unwatched status is lost from the pos file. (`FFFF...` is overwritten by the new position.)
* In this situation, the TailWatcher works correctly until the next rotation.
* When the next rotation occurs, `update_watcher` is wrongly skipped because the PositionEntry is already deleted from the run-time data (`PositionFile::map`).
  * https://github.com/fluent/fluentd/blob/78c91468d65da7cce0999fd94dbe61f7a35b606b/lib/fluent/plugin/in_tail.rb#L526-L529
  * (Fluentd v1.15.1 (td-agent v4.4.0) and later) The WARN log `Skip update_watcher because watcher has been already updated by other inotify event` occurs. (Fluentd v1.15.1 and later)
    * Before that version, it was DEBUG level.
* As a result, the TailWatcher with the old file handler continues working, and the new log file is not tailed.

The feature of this problem.

* Multiple position entries for the same path exist in the pos file.
* Log: `Skip update_watcher because watcher has been already updated by other inotify event`
* Log: `... already exists. use latest one: deleted #<...>` (When restarting Fluentd)

Workaround:

* `follow_inodes true` (Fluentd v1.16.2 (fluent-package v5.0.0) or later)
* `rotate_wait 0`

**Comment by daipom** [comment]
> Thanks for the update @daipom. pos file seems to be ok. `/var/log/service.log 000000000097d9f3 00000000002f710b`

Sorry for being late.
I currently think the above is the cause of this problem.

I'm not sure why your pos file has nothing wrong.

After restarting Fluentd, the pos file will recover with the log `... already exists. use latest one: deleted #<...>`.
I'm guessing that you restarted Fluentd after this problem and the file was recovered.

If there seems to be a different phenomenon going on than I assumed, please let me know.

**Comment by artheus** [comment]
My suggestion to fix this would be to add a native component to fluentd for
handling the in_tail properly. Using the in_tail code for fluent-bit, to
make a native component would probably not be much work to get it working
properly.

It could be built in a separate project, which would allow for the build of
fluentd to just download the binaries as a dependency.

From what I can gather, from struggling with this issue for a while now,
this would be the safest solution to this problem.

**Comment by artheus** [comment]
I've also made a small tool in Golang for easily converting the pos files
into the SQLLite db used by fluent-bit.

I'll look into sharing this tool here, as soon as I have the opportunity to
do so.

**Comment by artheus** [comment]
Here's the pos file migration script I wrote. (Seems I forgot that I just wrote it in bash, and not Golang)

It requires is `sqlite3` cli tool to be installed, for the script to work. Installation guides can be found here: https://www.tutorialspoint.com/sqlite/sqlite_installation.htm

```bash file="migrate.sh"
#!/usr/bin/env bash

info() {
  printf "[%s] INFO - %s\n" "$(date --iso-8601=seconds )" "$@"
}

readonly DB='/opt/fluent-bit-db/log-tracking.db'
readonly FLUENTD_LOG_POS="/var/log/fluentd-containers.log.pos"

if [[ ! -f "$FLUENTD_LOG_POS" ]]; then
  info "No FluentD log tracking file to migrate from"
  exit
fi

if [[ ! -f "$DB" ]]; then
  sqlite3 "$DB" "CREATE TABLE main.in_tail_files (id INTEGER PRIMARY KEY, name TEXT, offset INTEGER, inode INTEGER, created INTEGER, rotated INTEGER);"
else
  info "fluent-bit database already exists, will not do migration"
  exit
fi

while read -r line; do
  IFS=$'\t' read -r -a parts <<< "$line"

  filename="${parts[0]}"
  offset="$((16#${parts[1]}))"
  inode="$((16#${parts[2]}))"
  now="$(date +%s)"

  sqlite3 "$DB" "INSERT INTO in_tail_files (name, offset, inode, created, rotated) VALUES ('$filename', $offset, $inode, $now, 0)"
done < <(sort "$FLUENTD_LOG_POS")
```

There are no security stuff like escaping values for `INSERT INTO` or anything, but it's intended usage is in controlled environments. I am guessing there are plenty of things to improve on, but this has worked well on our internal k8s testing nodes, where we've used it as an init container for the fluent-bit pod which takes over for Fluentd.

**Comment by inc5025** [comment]
> Workaround:
> - follow_inodes true (Fluentd v1.16.2 (fluent-package v5.0.0) or later)
> - rotate_wait 0

I know that `rotate_wait` feature helps prevent data designated for the old file from getting lost. It seems that this doesn't work if `rotate_wait 0` is set, is there any side effect of workaround?

**Comment by daipom** [comment]
Thank you all very much!
Thanks to all your help, we were able to identify the logic problems and fix them!

The problem that `in_tail` stops collecting is fixed at Fluentd v1.16.3, Fluent Package v5.0.2 and td-agent v4.5.2.
Please use them!
If you still have problems, we'd be very grateful if you report to us.

**Comment by alexei-g-aloteq** [comment]
No, it is not fixed.
We've switched to image [1.16.3-debian-elasticsearch8-amd64-1.0](https://hub.docker.com/layers/fluent/fluentd-kubernetes-daemonset/v1.16.3-debian-elasticsearch8-amd64-1.0/images/sha256-ba8e2974dda9284ddf861c504ec52b150cfc0b6ab67246317d4b5373329895c3?context=explore), but still see almost daily outage.

**Comment by uristernik** [comment]
We are still seeing this issue as well. We are using this image [v1.16.3-debian-forward-1.0](https://github.com/fluent/fluentd-kubernetes-daemonset#:~:text=debian%2Dcloudwatch%2D1-,Forward,-docker%20pull%20fluent).

We were seeing the following message `If you keep getting this message, please restart Fluentd` and like suggested here we changed `follow_inodes` to `true` and set `rotate_wait` to `0`, but we are still seeing loads of `Skip update_watcher because watcher has been already updated by other inotify event`

We also noticed a pattern of memory leaking and gradual increase in CPU usage until a restart occurs. We are using fluentd as a daemonset on a kubernetes cluster. Here is our `in_tail` config:

```
<source>
  @type tail
  @id in_tail_container_logs
  path /var/log/containers/*.log
  pos_file /var/log/fluentd-containers.log.pos
  tag kubernetes.*
  read_from_head true
  follow_inodes true
  rotate_wait 0
  exclude_path ["/var/log/containers/fluentd*.log"]
  <parse>
    @type multi_format
    <pattern>
      format json
      time_key time
      time_type string
      time_format "%Y-%m-%dT%H:%M:%S.%NZ"
      keep_time_key true
    </pattern>
    
    <pattern>
      format /^(?<time>.+?) (?<stream>stdout|stderr) (?<logtag>[FP]) (?<log>.+)$/
      time_format "%Y-%m-%dT%H:%M:%S.%N%:z"
    </pattern>
    
  </parse>
  emit_unmatched_lines true
</source>

<filter kubernetes.**>
  @type concat
  key log
  partial_key logtag
  partial_value P
  separator ""
</filter>
```

**Comment by uristernik** [comment]
> We are still seeing this issue as well. We are using this image [v1.16.3-debian-forward-1.0](https://github.com/fluent/fluentd-kubernetes-daemonset#:~:text=debian%2Dcloudwatch%2D1-,Forward,-docker%20pull%20fluent).
> 
> We were seeing the following message `If you keep getting this message, please restart Fluentd` and like suggested here we changed `follow_inodes` to `true` and set `rotate_wait` to `0`, but we are still seeing loads of `Skip update_watcher because watcher has been already updated by other inotify event`
> 
> We also noticed a pattern of memory leaking and gradual increase in CPU usage until a restart occurs. We are using fluentd as a daemonset on a kubernetes cluster. Here is our `in_tail` config:
> 
> ```
> <source>
>   @type tail
>   @id in_tail_container_logs
>   path /var/log/containers/*.log
>   pos_file /var/log/fluentd-containers.log.pos
>   tag kubernetes.*
>   read_from_head true
>   follow_inodes true
>   rotate_wait 0
>   exclude_path ["/var/log/containers/fluentd*.log"]
>   <parse>
>     @type multi_format
>     <pattern>
>       format json
>       time_key time
>       time_type string
>       time_format "%Y-%m-%dT%H:%M:%S.%NZ"
>       keep_time_key true
>     </pattern>
>     
>     <pattern>
>       format /^(?<time>.+?) (?<stream>stdout|stderr) (?<logtag>[FP]) (?<log>.+)$/
>       time_format "%Y-%m-%dT%H:%M:%S.%N%:z"
>     </pattern>
>     
>   </parse>
>   emit_unmatched_lines true
> </source>
> 
> <filter kubernetes.**>
>   @type concat
>   key log
>   partial_key logtag
>   partial_value P
>   separator ""
> </filter>
> ```

In an environment where we have high volatility (we constantly deploy new code -> deployments are restarted very frequently -> pods are created -> lots of files to tail) we see a clear leak pattern.
Like mentioned above this is coupled with a huge amount of:
```
 Skip update_watcher because watcher has been already updated by other inotify event path="/var/log/containers/...
```
![CleanShot 2024-01-01 at 09 35 06@2x](https://github.com/fluent/fluentd/assets/39219232/3b94b3b6-17b7-4372-876f-5b5dd8bb71c1)


Any suggestion to mitigate this? @ashie @daipom 
Maybe setting `pos_file_compaction_interval` to some value so we would stop watching rotated files? (of pods that no longer exist)

**Comment by alexei-g-aloteq** [comment]
BTW, any good alternatives for Fluentd? 
It does not seem to be reliable at the moment. 
This issue kept unfixed for years.

**Comment by zmedico** [comment]
I've observed this issue with v1.16.3. I suspect that in_tail handling of file rotations is unlikely to ever reach a satisfactory level of reliability, and something like the [docker fluentd logging driver](https://docs.docker.com/config/containers/logging/fluentd/) (which unfortunately breaks `kubectl logs`) is the only way to reliably avoid problems with log file rotation.

I suspect that my `open_on_every_update true` config might be making problems worse (not to mention that it introduces a race to collect log entries during log rotation), so I'm going to try disabling that to see if it helps.

**Comment by zmedico** [comment]
Setting `open_on_every_update false` did not seem to help for me. Interestingly, the issue did not appear to correlate with a log rotation, since the log file that it was not following was still older than the fluentd process. Fluentd was following the file through two levels of symlinks, and the symlink timestamps are older than the actual log file, so I don't know how to explain why it triggered the "unreadable" log message.

**Comment by daipom** [comment]
Thanks for reporting.

> We are still seeing this issue as well. We are using this image [v1.16.3-debian-forward-1.0](https://github.com/fluent/fluentd-kubernetes-daemonset#:~:text=debian%2Dcloudwatch%2D1-,Forward,-docker%20pull%20fluent).
> 
> We were seeing the following message `If you keep getting this message, please restart Fluentd` and like suggested here we changed `follow_inodes` to `true` and set `rotate_wait` to `0`, but we are still seeing loads of `Skip update_watcher because watcher has been already updated by other inotify event`
> 
> We also noticed a pattern of memory leaking and gradual increase in CPU usage until a restart occurs. We are using fluentd as a daemonset on a kubernetes cluster. Here is our `in_tail` config:
> 

So,  there are still problems with both `follow_inodes false` and `follow_inodes true`, such as collection stops and resource leaks.

**Comment by ashie** [comment]
#4326 would be a possible remaining cause of this issue.
On the other hand it might not be related with above additional reports because #4326 says no error log is found.

**Comment by ashie** [comment]
BTW it would be better to open a new issue to treat remaining things.
This issue is too long and mixed various causes.
It seems hard to collect & track informations to debug remaining causes in this issue.

**Comment by uristernik** [comment]
@ashie @daipom Thanks for the responses and attention.

Should I open a new issue for that?

If you have any suggestion I'll be more than happy to test different configs.

**Comment by uristernik** [comment]
@ashie @daipom Thanks for the responses and attention.

Should I open a new issue for that?

If you have any suggestion I'll be more than happy to test different configs.
I added the `pos_file_compaction_interval 20m` to the config but still seeing that Fluentd is leaking CPU 

```

<source>
  @type tail
  @id in_tail_container_logs
  path /var/log/containers/*.log
  pos_file /var/log/fluentd-containers.log.pos
  tag kubernetes.*
  read_from_head true
  follow_inodes true
  rotate_wait 0
  exclude_path ["/var/log/containers/fluentd*.log", "/var/log/containers/scaleops*.log", "/var/log/containers/wiz-sensor*.log", "/var/log/containers/*kube-system*.log", "/var/log/containers/datadog*.log", "/var/log/containers/*calico-system*.log", "/var/log/containers/prometheus-node-exporter*.log", "/var/log/containers/opentelemetry-agent*.log"]
  pos_file_compaction_interval 20m
  <parse>
    @type multi_format
    <pattern>
      format json
      time_key time
      time_type string
      time_format "%Y-%m-%dT%H:%M:%S.%NZ"
      keep_time_key true
    </pattern>
    <pattern>
      format /^(?<time>.+?) (?<stream>stdout|stderr) (?<logtag>[FP]) (?<log>.+)$/
      time_format "%Y-%m-%dT%H:%M:%S.%N%:z"
    </pattern>
  </parse>
  emit_unmatched_lines true
</source>
```

<img width="1585" alt="CleanShot 2024-01-07 at 09 07 02@2x" src="https://github.com/fluent/fluentd/assets/39219232/b5d7b1b0-4b6e-425b-a276-e5e6e9c77872">

**Comment by daipom** [comment]
@uristernik Thanks. Could you please open a new issue?

**Comment by uristernik** [comment]
@daipom Done, hopefully I described the issue clear enough. Please correct me if I wasn't accurate enough

**Comment by daipom** [comment]
If you all still have a similar issue, please report it on the new issue!
